src/mpi-base.jl

```diff
@@ -156,6 +120,54 @@ function Comm_size(comm::Comm)
     Int(size[])
 end
 
+function Type_create_struct{T <: Any}(::Type{T})  # <: Any effectively 
+                                                  # limits T to being a Type
+
+  @assert isbits(T)
+  # get the data from the type
+  fieldtypes = T.types
+  offsets = fieldoffsets(T)
+  nfields = Cint(length(fieldtypes))
+
+  # put data in MPI format
+  blocklengths = ones(Cint, nfields)
+  displacements = zeros(Cptrdiff_t, nfields)  # size_t == MPI_Aint ?
+  types = zeros(Cint, nfields)
+  for i=1:nfields
+    displacements[i] = offsets[i]
+    types[i] = mpitype(fieldtypes[i])
+  end
+
+  # create the datatype
+  newtype_ref = Ref{Cint}()
+  flag = Ref{Cint}()
+  ccall(MPI_TYPE_CREATE_STRUCT, Void, (Ptr{Cint}, Ptr{Cint}, Ptr{Cptrdiff_t}, 
+        Ptr{Cint}, Ptr{Cint}, Ptr{Cint}), &nfields, blocklengths, displacements, types, 
+        newtype_ref, flag)
+
+  if flag[] != 0
+    println(STDERR, "Warning: MPI_TYPE_CREATE_STRUCT returned non-zero exit stats")
```

Comment:
Outputting an error to stderr doesn't seem right; most likely, people will not see this. It's customary in Julia to call `throw` instead.